Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,330 @@
+public var _useLegacyIdentities: Bool = true {
+    willSet {
+        PackageIdentity.provider = newValue ? LegacyPackageIdentity.self : CanonicalPackageIdentity.self
+    }
+}
+
+internal protocol PackageIdentityProvider: LosslessStringConvertible {
+    init(_ string: String)
+}
+
+/// The canonical identifier for a package, based on its source location.
+public struct PackageIdentity: LosslessStringConvertible, Hashable {
+    internal static var provider: PackageIdentityProvider.Type = LegacyPackageIdentity.self
+
+    /// A textual representation of this instance.
+    public let description: String
+
+    /// Instantiates an instance of the conforming type from a string representation.
+    public init(_ string: String) {
+        self.description = Self.provider.init(string).description
+    }
+}
+
+extension PackageIdentity: Comparable {
+    public static func < (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.description < rhs.description
+    }
+}
+
+extension PackageIdentity: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let description = try container.decode(String.self)
+        self.init(description)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.description)
+    }
+}
+
+extension PackageIdentity: JSONMappable, JSONSerializable {
+    public init(json: JSON) throws {
+        guard case .string(let string) = json else {
+            throw JSON.MapError.typeMismatch(key: "", expected: String.self, json: json)
+        }
+
+        self.init(string)
+    }
+
+    public func toJSON() -> JSON {
+        return .string(self.description)
+    }
+}
+
+// MARK: -
+
+struct LegacyPackageIdentity: PackageIdentityProvider, Equatable {
+    /// A textual representation of this instance.
+    public let description: String
+
+    /// Instantiates an instance of the conforming type from a string representation.
+    public init(_ string: String) {
+        self.description = PackageReference.computeDefaultName(fromURL: string).lowercased()
+    }
+}
+
+/// A canonicalized package identity.
+///
+/// A package may declare external packages as dependencies in its manifest.
+/// Each external package is uniquely identified by the location of its source code.
+///
+/// An external package dependency may itself have one or more external package dependencies,
+/// known as _transitive dependencies_.
+/// When multiple packages have dependencies in common,
+/// Swift Package Manager determines which version of that package should be used
+/// (if any exist that satisfy all specified requirements)
+/// in a process called package resolution.
+///
+/// External package dependencies are located by a URL
+/// (which may be an implicit `file://` URL in the form of a file path).
```

Comment:
Should this description also mention path-based external dependencies?  They do not use `file://`, which is only used when the local path contains an actual local Git repository and not just an unversioned package.  Or is package identity only a thing for actual remote packages (not root or local)?

Reply:
That's what I was trying to get at — my conceit here is that a file path is _actually_ a `file:///` URL with an implicit scheme. Any package dependency, local or remote, is described by a URL: local use `file:///` URLs and remote use any other scheme. `CanonicalPackageIdentity` distinguishes between local from remote by preserving a leading slash (for example, you couldn't shadow a remote URL by having a root directory named `/github.com/`). Does that all make sense?

Reply:
Understood, but I think that's too subtle and confusing of a way to represent it.  And checking the file system for whether a particular path has a .git isn't a good way to do it either.  I think there should be a clear way to represent that a package is either an unversioned take-it-as-given path in the file system or a repository that just happens to be in the local file system.  Historically there has been value in using no scheme for the former and file: for the latter.  I think it would be a mistake to conflate them, even it's through an extra slash character.